src/common/store/gameModes.js

```diff
@@ -0,0 +1,67 @@
+import { cloneDeep, shuffle } from 'lodash';
+import seededRNG from 'seed-random';
+
+import { triggerSound } from '../util/game';
+
+import defaultState, { bluePlayerState, orangePlayerState } from './defaultGameState';
+
+export class GameMode {
+  name = undefined;
+
+  static fromString(gameModeStr) {
+    const modes = [ NormalGameMode, SharedDeckGameMode ];
+    const mode = modes.find(m => m.name === gameModeStr);
+    if (!mode) {
+      throw `Unknown game mode: ${gameModeStr}`;
+    }
+    return mode;
+  }
+
+  isActive(state) {
+    return state.gameMode === this.name;
+  }
+
+  startGame(state, player, usernames, decks, seed) {
+    state = Object.assign(state, cloneDeep(defaultState), {
+      gameMode: this.name,
+      player: player,
+      rng: seededRNG(seed),
+      started: true,
+      usernames
+    });
+    state = triggerSound(state, 'yourmove.wav');
+
+    return state;
+  }
+}
+
+export const NormalGameMode = new (class extends GameMode {
```

Comment:
Woah this is some pretty weird syntax. Is this an anonymous class?

Reply:
Why can't this be `export class NormalGameMode extends GameMode {`?

Reply:
Alas, that seems to be the least objectionable way to make singleton objects in ES6.